audittools/backing_store_file.go

```diff
@@ -0,0 +1,460 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company
+// SPDX-License-Identifier: Apache-2.0
+
+package audittools
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/majewsky/gg/option"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-api-declarations/cadf"
+
+	"github.com/sapcc/go-bits/logg"
+)
+
+// NewFileBackingStore creates a file-based backing store from JSON parameters.
+// This is the factory function for use in AuditorOpts.BackingStoreFactories.
+//
+// Example usage:
+//
+//	auditor, err := NewAuditor(ctx, AuditorOpts{
+//	    BackingStoreFactories: map[string]BackingStoreFactory{
+//	        "file": NewFileBackingStore,
+//	    },
+//	})
+func NewFileBackingStore(params json.RawMessage, opts AuditorOpts) (BackingStore, error) {
+	var store FileBackingStore
+	if err := json.Unmarshal(params, &store); err != nil {
+		return nil, fmt.Errorf("audittools: failed to parse file backing store config: %w", err)
+	}
+
+	registry := opts.Registry
+	if registry == nil {
+		registry = prometheus.DefaultRegisterer
+	}
+
+	if err := store.Init(registry); err != nil {
+		return nil, err
+	}
+	return &store, nil
+}
+
+// FileBackingStore implements BackingStore using local filesystem files.
+// Provides durable audit buffering for services with persistent volumes.
+//
+// Thread safety: Write() and ReadBatch() are serialized by a mutex.
+// Multiple concurrent calls are safe but will block each other.
+// Callers must ensure that commit() completes before the next ReadBatch() call.
+type FileBackingStore struct {
+	// Configuration (JSON params)
+	Directory    string               `json:"directory"`
+	MaxFileSize  option.Option[int64] `json:"max_file_size"`
+	MaxTotalSize option.Option[int64] `json:"max_total_size"`
```

Comment:
Having `Option` here means that each method of FileBackingStore needs to handle the None case, even though you are removing the None case during Init. I suggest having just `int64` here, and only having Option at params parsing time, i.e. in NewFileBackingStore:

```go
var cfg struct {
	Directory    string               `json:"directory"`
	MaxFileSize  option.Option[int64] `json:"max_file_size"`
	MaxTotalSize option.Option[int64] `json:"max_total_size"`
}
if err := json.Unmarshal(params, &cfg); err != nil {
	return nil, fmt.Errorf("audittools: failed to parse file backing store config: %w", err)
}
store := FileBackingStore {
	Directory:    cfg.Directory,
	MaxFileSize:  cfg.MaxFileSize.UnwrapOr(10<<20),
	MaxTotalSize: cfg.MaxTotalSize.UnwrapOr(0),
}
```